### ISSUE TITLE
Diff not generated for protected resource which needs to be replaced

### ISSUE THREAD

**Issue by nesl247** [pulumi/pulumi#5749]
When a protected resource needs to be replaced, there is no diff available to see why it needs to be replaced. It only generates an error stating it cannot replace a protected resource.

## Expected Behavior
The diff should still be generated.

## Current Behavior
```
error: Preview failed: refusing to delete protected resource 'urn:pulumi:staging::shop-front::gcp:redis/instance:Instance::shop-front-ha'
```

## Steps to Reproduce
1. Create any resource as protected
2. Make a change to the resource that would require it to be replaced
3. Try to run `pulumi up`
4. Error

## Context (Environment)
In this case we are trying to deploy a change to k8s which has a redis instance in GCP in this project. I don't know what the cause for the replacement is because of this issue.

**Comment by wilsoncg** [comment]
+1
I am experiencing the same bug, very confusing. Preview (& up) fails, reports that old state & new state are the same, yet wishes to perform a replacement on a protected resource & fails with no diff generated.

This is a fresh import into a stack, with a newly minted corresponding C# project. I would expect pulumi to report there is nothing to do for a 'same' operation. Why is it failing?

How should I proceed?

```
PS> pulumi preview --json
```
```javascript
{
    "config": {
        "azure:location": "WestEurope"
    },
    "steps": [
        {
            "op": "same",
            "urn": "urn:pulumi:bot::bot::pulumi:pulumi:Stack::bot-bot",
            "oldState": {
                "urn": "urn:pulumi:bot::bot::pulumi:pulumi:Stack::bot-bot",
                "custom": false,
                "type": "pulumi:pulumi:Stack"
            },
            "newState": {
                "urn": "urn:pulumi:bot::bot::pulumi:pulumi:Stack::bot-bot",
                "custom": false,
                "type": "pulumi:pulumi:Stack"
            },
            "detailedDiff": null
        }
    ],
    "diagnostics": [
        {
            "urn": "urn:pulumi:bot::bot::azure-nextgen:alertsmanagement/latest:SmartDetectorAlertRule::Failure Anomalies - RachaelBot4wwyzn",
            "message": "error: Preview failed: refusing to delete protected resource 'urn:pulumi:bot::bot::azure-nextgen:alertsmanagement/latest:SmartDetectorAlertRule::Failure Anomalies - RachaelBot4wwyzn'\n",
            "severity": "error"
        },
        {
            "message": "error: preview failed\n",
            "severity": "error"
        }
    ],
    "duration": 3543236500,
    "changeSummary": {
        "same": 4
    }
}
```

```
PS> pulumi preview --show-replacement-steps -v 3
Previewing update (bot)

View Live: https://app.pulumi.com/wilsoncg/bot/bot/previews/c75fe0fb-9b27-4d6a-a068-4a7e53a694e0

     Type                                                             Name                                  Plan        Info
     pulumi:pulumi:Stack                                              bot-bot                                           1 error
 +-  └─ azure-nextgen:alertsmanagement/latest:SmartDetectorAlertRule  Failure Anomalies - RachaelBot4wwyzn  replace     1 error

Diagnostics:
  azure-nextgen:alertsmanagement/latest:SmartDetectorAlertRule (Failure Anomalies - RachaelBot4wwyzn):
    error: Preview failed: refusing to delete protected resource 'urn:pulumi:bot::bot::azure-nextgen:alertsmanagement/latest:SmartDetectorAlertRule::Failure Anomalies - RachaelBot4wwyzn'
 
  pulumi:pulumi:Stack (bot-bot):
    error: preview failed
```

**Comment by gpduck** [comment]
This is important for us as well. More often than not we want to correct the script/environment to remove the need for the replacement, but we are not given any clues as to what the cause is so it makes troubleshooting very difficult.

**Comment by kaarelr** [comment]
+1

This is crucial when debugging unexpected resource replacements.

**Comment by dixler** [comment]
@kaarelr might you be running into
- https://github.com/pulumi/pulumi/issues/12964

**Comment by sam-utila** [comment]
Any update on this?

**Comment by joeduffy** [comment]
@Frassle Do you see any reasons we can't just move the check here into the actual step execution, rather than step generation? I created a small patch that did this and it behaves as I would hope -- allowing you to see the diff. The only risk I see is whether we try to perform the replacement and let the delete go through (depending on step order) but that doesn't seem to be the case -- the replace step is executed in the correct dependency order.

```patch
diff --git a/pkg/resource/deploy/step.go b/pkg/resource/deploy/step.go
index f261178e0..f6bda2a91 100644
--- a/pkg/resource/deploy/step.go
+++ b/pkg/resource/deploy/step.go
@@ -617,6 +617,18 @@ func (s *ReplaceStep) Apply(preview bool) (resource.Status, StepCompleteFunc, er
        // If this is a pending delete, we should have marked the old resource for deletion in the CreateReplacement step.
        contract.Assertf(!s.pendingDelete || s.old.Delete,
                "old resource %v should be marked for deletion if pending delete", s.old.URN)
+
+       // If this resource is protected we can't replace it because that entails a delete
+       // Note that we do allow unprotecting and replacing to happen in a single update
+       // cycle, we don't look at old.Protect here.
+       if s.new.Protect && s.old.Protect {
+               message := fmt.Sprintf("unable to replace resource %q\n"+
+                       "as it is currently marked for protection. "+
+                       "To unprotect the resource, remove the `protect` flag from the resource in your Pulumi "+
+                       "program and run `pulumi up`", s.old.URN)
+               return resource.StatusOK, func() {}, errors.New(message)
+       }
+
        return resource.StatusOK, func() {}, nil
 }

diff --git a/pkg/resource/deploy/step_generator.go b/pkg/resource/deploy/step_generator.go
index d7bc28963..e2a6be634 100644
--- a/pkg/resource/deploy/step_generator.go
+++ b/pkg/resource/deploy/step_generator.go
@@ -1024,19 +1024,6 @@ func (sg *stepGenerator) generateStepsFromDiff(
        // If there were changes check for a replacement vs. an in-place update.
        if diff.Changes == plugin.DiffSome {
                if diff.Replace() {
-                       // If this resource is protected we can't replace it because that entails a delete
-                       // Note that we do allow unprotecting and replacing to happen in a single update
-                       // cycle, we don't look at old.Protect here.
-                       if new.Protect && old.Protect {
-                               message := fmt.Sprintf("unable to replace resource %q\n"+
-                                       "as it is currently marked for protection. To unprotect the resource, "+
-                                       "remove the `protect` flag from the resource in your Pulumi "+
-                                       "program and run `pulumi up`", urn)
-                               sg.deployment.ctx.Diag.Errorf(diag.StreamMessage(urn, message, 0))
-                               sg.sawError = true
-                               return nil, result.BailErrorf(message)
-                       }
-
                        // If the goal state specified an ID, issue an error: the replacement will change the ID, and is
                        // therefore incompatible with the goal state.
                        if goal.ID != "" {
```

**Comment by Frassle** [comment]
> Do you see any reasons we can't just move the check here into the actual step execution

If we're doing a create before delete replace the create will still happen before that delete step fails. We could have a check in the step gen to create an "ErroringReplaceStep" maybe so it still makes a step for the diff view but always just fails.

**Comment by mikhailshilkov** [comment]
I believe this is another instance of https://github.com/pulumi/pulumi/issues/5027 and has been fixed in https://github.com/pulumi/pulumi/pull/15969 and released in the CLI 3.113.2
